### ISSUE TITLE
"infohash merge conflict" leads to removing handles without `torrent_removed_alert`

### ISSUE THREAD

**Issue by AllSeeingEyeTolledEweSew** [arvidn/libtorrent#6917]
libtorrent version (or branch): `libtorrent==2.0.6` from pypi

platform/architecture: Linux x86_64

Consider a v1-only magnet link, and a v2-only magnet link, that actually refer to the hashes of the same hybrid torrent. These magnet links may be added to a session, and libtorrent will create separate handles for them. When either torrent receives metadata, libtorrent realizes it conflicts with the other torrent, and puts it in an error state. I call this an "infohash merge conflict".

This is tricky. Many different behaviors by libtorrent might be valid. I couldn't find documentation on what to expect. I'll file a separate issue to narrow down what guarantees libtorrent should provide, and to document these.

This issue is about a specific issue with this case that is definitely wrong: if both handles are in a "merge conflict" error state, then removing either one from the session implicitly removes the other one, but `torrent_removed_alert` is only fired for one. This will definitely lead to bad state in the app.

Here's a python script that generates this case. I create a merge conflict, test some properties of each handle, then try to remove them both. Note that removing one handle invalidates the other one, and `torrent_removed_alert` is never fired for the second handle.

```python
import libtorrent as lt
import sys
import traceback
import requests


def main() -> int:
    # Use the official hybrid torrent
    resp = requests.get("https://libtorrent.org/bittorrent-v2-hybrid-test.torrent")
    ti = lt.torrent_info(lt.bdecode(resp.content))
    # Create separate magnet links for the v1 and v2 info hashes
    atp1 = lt.parse_magnet_uri(f"magnet:?xt=urn:btih:{ti.info_hashes().v1}")
    atp2 = lt.parse_magnet_uri(f"magnet:?xt=urn:btmh:1220{ti.info_hashes().v2}")

    # Add the magnet links as separate torrent_handles
    session = lt.session({"alert_mask": lt.alert_category.status})
    handle1 = session.add_torrent(atp1)
    handle2 = session.add_torrent(atp2)

    # Inject the metadata (rather than waiting for ut_metadata)
    handle1.set_metadata(ti.info_section())
    handle2.set_metadata(ti.info_section())

    def ih_str(ih: lt.info_hash_t) -> str:
        return f"[{ih.v1},{ih.v2}]"

    # Print some features about each handle
    print("after set_metadata(), ...")
    print(f"handle1 has info_hashes(): {ih_str(handle1.info_hashes())}")
    print(f"handle2 has info_hashes(): {ih_str(handle2.info_hashes())}")
    print(f"handle1 has torrent_file(): {handle1.torrent_file()}")
    print(f"handle2 has torrent_file(): {handle2.torrent_file()}")
    print()

    # Experiment with whether save_resume_data() works
    handle1.save_resume_data()
    handle2.save_resume_data()

    # Print alerts from the session
    while True:
        session.wait_for_alert(1000)
        for alert in session.pop_alerts():
            # Ensure we can distinguish which handle generated each alert
            if isinstance(alert, lt.torrent_alert):
                matches: list[str] = []
                if alert.handle == handle1:
                    matches.append("v1")
                if alert.handle == handle2:
                    matches.append("v2")
                which = ",".join(matches)
                print(f"{alert.__class__.__name__}({which}): {alert.message()}")
            else:
                print(f"{alert.__class__.__name__}: {alert.message()}")
            if isinstance(alert, (lt.save_resume_data_alert, lt.save_resume_data_failed_alert)):
                print("removing handle...")
                try:
                    session.remove_torrent(alert.handle)
                except Exception as exc:
                    traceback.print_exc(file=sys.stdout)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
```
```
after set_metadata(), ...
handle1 has info_hashes(): [631a31dd0a46257d5078c0dee4e66e26f73e42ac,0000000000000000000000000000000000000000000000000000000000000000]
handle2 has info_hashes(): [0000000000000000000000000000000000000000,d8dd32ac93357c368556af3ac1d95c9d76bd0dff6fa9833ecdac3d53134efabb]
handle1 has torrent_file(): <libtorrent.torrent_info object at 0x7a1fca2ae6d0>
handle2 has torrent_file(): <libtorrent.torrent_info object at 0x7a1fca2ae7b0>

listen_succeeded_alert: successfully listening on [TCP] 127.0.0.1:6881
listen_succeeded_alert: successfully listening on [uTP] 127.0.0.1:6881
listen_succeeded_alert: successfully listening on [TCP] [::1]:6881
listen_succeeded_alert: successfully listening on [uTP] [::1]:6881
add_torrent_alert(v1): added torrent: 631a31dd0a46257d5078c0dee4e66e26f73e42ac
torrent_added_alert(v1): 631a31dd0a46257d5078c0dee4e66e26f73e42ac added
state_changed_alert(v1): 631a31dd0a46257d5078c0dee4e66e26f73e42ac: state changed to: dl metadata
add_torrent_alert(v2): added torrent: d8dd32ac93357c368556af3ac1d95c9d76bd0dff
torrent_added_alert(v2): d8dd32ac93357c368556af3ac1d95c9d76bd0dff6fa9833ecdac3d53134efabb added
state_changed_alert(v2): d8dd32ac93357c368556af3ac1d95c9d76bd0dff6fa9833ecdac3d53134efabb: state changed to: dl metadata
torrent_error_alert(v1): bittorrent-v1-v2-hybrid-test ERROR: (19 torrent already exists in session) metadata
torrent_error_alert(v2): bittorrent-v1-v2-hybrid-test ERROR: (19 torrent already exists in session) metadata
save_resume_data_failed_alert(v1): bittorrent-v1-v2-hybrid-test resume data was not generated: torrent removed
removing handle...
save_resume_data_failed_alert(v2): bittorrent-v1-v2-hybrid-test resume data was not generated: torrent removed
removing handle...
Traceback (most recent call last):
  File "/home/amnesia/Persistent/checkout/tvaf/t.py", line 50, in main
    session.remove_torrent(alert.handle)
RuntimeError: invalid torrent handle used
torrent_removed_alert(v1): bittorrent-v1-v2-hybrid-test removed
```

**Comment by arvidn** [comment]
I believe this is fully resolved, right?

**Comment by AllSeeingEyeTolledEweSew** [comment]
yep, fixed in 2.0.7
